### DISCUSSION TITLE
What is the intended behavior of Cloud Sync?

### DISCUSSION THREAD

**Discussion by RhythmScript** [mtkennerly/ludusavi#436] [Q&A]
tl;dr
**Is the intended behavior of cloud sync just to create a redundant back up? Is there no way of having ludusavi restore from the cloud path TO the local path as part of the restore process?**

I've been working for the last several days on getting automatic cross-platform save synchronization set up between my Steam Deck and my Windows desktop using ludusavi. For the most part, I have gotten this set up to work well; I push data using the file structure on my Steam Deck into Dropbox. Dropbox then appears as a folder on Windows, where I can pull the data from using bidirectional redirects to unmap the linux/Proton/etc paths back into the desired locations on my Windows desktop.

When I originally set this up, I used the cloud saves functionality in ludusavi to push my deck backups to my Dropbox via rclone. I encountered a problem with bidirectional communication with this method, however.

Suppose I recover a save on Deck, modify it, and then back it up again on game close using a launch wrapper. That works fine.
Now, on my desktop PC, I can open the same game using a launch wrapper, and it will recover the cloud save, let me modify it, and then back it up again on close.
Now, when I reopen the save on Deck, we have an issue. See, the Deck _copies_ its local backup structure to the cloud, but it only RECOVERS from its _local_ recovery structure. It will automatically attempt a cloud sync _upload_ when it backs up, but it won't attempt a cloud sync _download_ when it recovers.

So, now the saves are out of sync. And they will stay that way, because now, when the Deck further modifies the save and backs it up, it will detect that the local backup and cloud backup are no longer identical at the start of backup (one has the last Deck local save at the time of the last backup, the other has the save that was modified on Windows instead), and it will cancel the cloud sync until those files are manually resynchronized. So my local save is there in the local path, but not the cloud; and if I recover again on Windows, I will have reverted to the first original Deck save and lose progress.

So, I think to myself, the solution here is to replicate the Windows behavior on Deck. Restoring from-cloud works on Windows because the cloud saves are an actual recovery path, not a separate cloud backup. So I configure rclone to mount my Dropbox as a path on the Deck's filesystem itself. This takes a bit of doing with a startup service to ensure it's always active, but it does work. I can now recover directly from the cloud path, rather than the local path.

And this works... but then it doesn't. Because now the Steam Deck can't _save_ locally.

See, the save that is recovered from the cloud is pushed directly to the main save location for the game in question. But it's still not pushed to the ludusavi local save backup location. So when ludusavi goes to perform the back-up from the local location to the cloud... it once again detects that the files are not in sync. And refuses to push the local save to the cloud save. The local save backup and cloud save backup are still desynced, because ludusavi can only PUSH files to the cloud automatically via rclone, it doesn't PULL them.

The final, most obvious fix is to just use the mount path for both the back up and recover location. And this does, indeed, "work". The problem is, rclone as a mount point is _significantly_ slower than using sync, since rclone mounts cannot make use of --fast-list. It takes on the order of 20 seconds to complete the backup of a single game save on game exit. And crucially, during this time, ludusavi does not lock Steam in any way or self-check during this time. So if you restart the game before that 20 seconds is up, you can wind up re-copying the existing save to the local path from the cloud, and permanently erasing what you were trying to push TO the cloud. This isn't an issue when the backup happens on the order of a second or less and it's using local caching but is a huge problem in this context.

So what I am asking, after all this backstory, is the following:
**Is the intended behavior of cloud sync just to create a redundant back up? Is there no way of having ludusavi restore from the cloud path TO the local path as part of the restore process?**

Because if you can only restore from the cloud manually, that locks you into using mountpoints for bidirectional cloud sync, and rclone simply is not optimized to handle mountpoints as well as it's optimized to handle regular data duplication in either direction.

**Comment by mtkennerly** [comment]
Hi! This is definitely a shortcoming of the cloud functionality as it stands today.

I would say the main purpose is to serve as a redundant backup, but it is also meant to support cross-system sync, although it's not as streamlined as it probably should be. Right now, when switching systems like this, you're just supposed to open Ludusavi and perform a download on the "other" screen manually, which I understand isn't so convenient. Since you're using `wrap`, probably that command should detect the cloud desync and ask what you want to do, like what Steam Cloud does when you launch a game. I can look into that.

For what you're trying to do, I know some people recommend [Syncthing](https://syncthing.net), although I haven't used it myself.

**Reply by Kirkland-Pickles** [reply]
> Hi! This is definitely a shortcoming of the cloud functionality as it stands today.
> 
> I would say the main purpose is to serve as a redundant backup, but it is also meant to support cross-system sync, although it's not as streamlined as it probably should be. Right now, when switching systems like this, you're just supposed to open Ludusavi and perform a download on the "other" screen manually, which I understand isn't so convenient. Since you're using `wrap`, probably that command should detect the cloud desync and ask what you want to do, like what Steam Cloud does when you launch a game. I can look into that.
> 
> For what you're trying to do, I know some people recommend [Syncthing](https://syncthing.net), although I haven't used it myself.

why not just add an option to automatically check that the cloud is aligned with the local files when the app launches? 

im attempting to set up cross platform sync now with a windows pc and ludusavi on a steam deck

the process is like this 

play game on windows pc - manually backup game save in ludusavi, which uploads it to cloud

switch to steam deck to continue playing with cloud sync 

i now must switch to desktop mode, and sync the cloud files back down to the steam decks local files, then switch back to game mode, open ludusavi-decky plugin, restore the save from there. 

play the game on the steam deck - automatically (yay) backup the save on game close (due to ludusavi-decky)

back to windows pc - must first open ludusavi app, download the cloud files to local, manually restore, play game. continuing where i left off. 


what am i missing in this process? is this really the smoothest ludusavi save sync can get?

**Comment by jbbandos** [comment]
I had a similar issue, but with steamdeck<->desktop linux. The solution for me was indeed syncthing, set up to sync the folder above the ludusavi backup folder. It was simple to set that up as I was already replicating my documents folder.
My setup now is ludusavi on all machines I need to sync game saves, using the folder "$HOME/ludusavi-backup/saves"  (eg "/home/deck/ludusavi-backup/saves" on the steamdeck) for the "backup to" and "restore from", backing up to dropbox, and translating the save locations, that is, using redirects on the non-steamdeck computers to change from steamdeck paths to my paths. I tend to have a separate partition/drive with my steam and heroic libraries, /media/Games, so it is easier - <img width="1318" height="203" alt="Screenshot_20260620_075146" src="https://github.com/user-attachments/assets/7bdfc928-c637-45a4-8820-8958ef7b80df" />

The syncthing folder configuration is very simple - 
<img width="899" height="675" alt="Screenshot_20260620_075857" src="https://github.com/user-attachments/assets/e4f06354-539a-4c23-bc82-9498ac0d6b75" />

Remember to do a download of the cloud files before adding the folder to syncthing, just to be sure you start with a known state,

I still have to find a way to do these automatic syncs when using multiple steam libraries - the redirects as I wrote them won't work there, no idea on how to handle it, how to restore a backup that was done for a few games in the same partition to another system where they are in separate paths.
